plugins/backstage-plugin-coder/src/api/CoderClient.ts

```diff
@@ -0,0 +1,348 @@
+    this.requestTimeoutMs = requestTimeoutMs;
+
+    this.cleanupController = new AbortController();
+    this.trackedEjectionIds = new Set();
+
+    this.sdk = this.getBackstageCoderSdk(this.axios);
+    this.addBaseRequestInterceptors();
+  }
+
+  private addRequestInterceptor(
+    requestInterceptor: (
+      config: RequestConfig,
+    ) => RequestConfig | Promise<RequestConfig>,
+    errorInterceptor?: (error: unknown) => unknown,
+  ): number {
+    const ejectionId = this.axios.interceptors.request.use(
+      requestInterceptor,
+      errorInterceptor,
+    );
+
+    this.trackedEjectionIds.add(ejectionId);
+    return ejectionId;
+  }
+
+  private removeRequestInterceptorById(ejectionId: number): boolean {
+    const sizeBeforeRemoval = this.trackedEjectionIds.size;
+
+    this.axios.interceptors.request.eject(ejectionId);
+    if (this.trackedEjectionIds.has(ejectionId)) {
+      this.trackedEjectionIds.delete(ejectionId);
+    }
+
+    return sizeBeforeRemoval !== this.trackedEjectionIds.size;
+  }
+
+  private addBaseRequestInterceptors(): void {
+    // Configs exist on a per-request basis; mutating the config for a new
+    // request won't mutate any configs for requests that are currently pending
+    const baseRequestInterceptor = async (
+      config: RequestConfig,
+    ): Promise<RequestConfig> => {
+      // Front-load the setup steps that rely on external APIs, so that if any
+      // fail, the request bails out early before modifying the config
+      const proxyApiEndpoint = await this.urlSync.getApiEndpoint();
+      const bearerToken = (await this.identityApi.getCredentials()).token;
+
+      config.baseURL = proxyApiEndpoint;
+      config.signal = this.getTimeoutAbortSignal();
+
+      // The Axios docs have incredibly confusing wording about how multiple
+      // interceptors work. They say the interceptors are "run in the order
+      // added", implying that the first interceptor you add will always run
+      // first. That is not true - they're run in reverse order, so the newer
+      // interceptors will always run before anything else. Only add token from
+      // this base interceptor if a newer interceptor hasn't already added one
+      if (config.headers[CODER_AUTH_HEADER_KEY] === undefined) {
+        config.headers[CODER_AUTH_HEADER_KEY] = this.loadedSessionToken;
+      }
+
+      if (bearerToken) {
+        config.headers.Authorization = `Bearer ${bearerToken}`;
+      }
+
+      return config;
+    };
+
+    const baseErrorInterceptor = (error: unknown): unknown => {
+      const errorIsFromCleanup =
+        error instanceof DOMException &&
+        error.name === sharedCleanupAbortReason.name &&
+        error.message === sharedCleanupAbortReason.message;
+
+      // Manually aborting a request is always treated as an error, even if we
+      // 100% expect it. Just scrub the error if it's from the cleanup
+      if (errorIsFromCleanup) {
+        return undefined;
+      }
```

Comment:
So, as far as I can tell from the Axios docs, the way the error interceptor works is that, Axios will run all the interceptors, and then if the value it gets back is not `undefined`, Axios will treat that value as an error and throw it (even if the value isn't an `Error` instance)

I could've just had `return`, but Backstage's linters complain about mixing `return` and `return <value>` in the same function

Reply:
Ah yup, I get that the `return undefined` here is causing Axios to not throw and not being used itself as the return of `client.getWorkspaces()`, but I wonder what now is the return of `client.getWorkspaces()` with the error having been scrubbed?  Since it did not throw, but there was also no data because it aborted, is it just naturally `undefined` maybe?

And if so, maybe returning undefined for canceled things is React-like so this is fine, was my thinking, but honestly I have no idea.  :stuck_out_tongue:  So you would write code like:

```
const workspaces = client.getWorkspaces()
if (workspaces === undefined) {
  // This means the request was canceled
}
```

Reply:
Ooh, yeah, good point – realistically, I only expect the `cleanupClient` function to be used for testing, but that is a hole in the logic that could pause pain points down the line

It's not the worst thing in the world – React Query exposes the `error` property as type `unknown` by default to be defensive, so technically, all RQ code already has to deal with an "error" potentially being undefined. But if I do need to keep this method for connecting to the Coder core files, I'll be sure to figure something out

Reply:
Oh yeah good point, this really only comes into play when cleaning up the client.